File: lib/ui_kit/components/ui/chip_input.ex
defmodule UiKit.Components.Ui.ChipInput do
  @moduledoc """
  Chip Input component for managing arrays as removable chips.

  Displays a list of values as chips (badges) with remove buttons, plus an input
  field for adding new values. Supports keyboard shortcuts and optional suggestions.

  ## Features

  - Display values as removable chips using Badge component
  - Input field for typing new values
  - Enter, Space, or Comma adds new chip
  - Backspace on empty input removes last chip
  - Click X button to remove specific chip
  - Optional suggestions dropdown while typing
  - Keyboard accessible navigation
  - Semantic token styling

  ## Basic Usage

      <.chip_input
        id="tags-input"
        values={@current_tags}
        placeholder="Add tag..."
        on_add={JS.push("add_tag")}
        on_remove={JS.push("remove_tag")}
      />

  ## With Suggestions

      <.chip_input
        id="tags-input"
        values={@current_tags}
        placeholder="Add tag..."
        suggestions={@available_tags}
        on_add="add_tag"
        on_remove="remove_tag"
      />

  ## LiveView Event Handling

      def handle_event("add_tag", %{"value" => tag}, socket) do
        {:noreply, update(socket, :current_tags, fn tags -> [tag | tags] end)}
      end

      def handle_event("remove_tag", %{"value" => tag}, socket) do
        {:noreply, update(socket, :current_tags, &List.delete(&1, tag))}
      end

  """
  use Phoenix.Component

  import UiKit.Components.CoreComponents, only: [icon: 1]
  import UiKit.Components.Ui.FeedbackStatus, only: [badge: 1]

  alias Phoenix.LiveView.JS
  alias Phoenix.LiveView.Rendered

  @doc """
  Renders a chip input component for managing array values.

  Values are displayed as badges with remove buttons. The input field allows
  adding new values via keyboard shortcuts (Enter, Space, Comma).

  ## Modes

  The component operates in two modes:

  ### Client-side mode (default)
  When `on_add` and `on_remove` are NOT provided, the component manages its own
  state entirely in JavaScript. Values are stored in a hidden input field with
  the name attribute, making it perfect for standard form submissions.

      <.chip_input
        id="tags-input"
        name="post[tags]"
        value={@post.tags}
      />

  ### Server-side mode
  When `on_add` or `on_remove` are provided, the component sends events to the
  server for every add/remove operation. Use this when you need server-side
  validation or real-time updates.

      <.chip_input
        id="tags-input"
        values={@tags}
        on_add="add_tag"
        on_remove="remove_tag"
      />

  ## Attributes

  - `id` - Required unique identifier
  - `name` - Form field name for client-side mode (e.g., "post[tags][]")
  - `value` - Initial values (client-side mode, can be list or comma-separated string)
  - `values` - Current values (server-side mode, list only)
  - `placeholder` - Placeholder text for input field (default: "Add value...")
  - `suggestions` - Optional list of suggested values to show in dropdown
  - `on_add` - Event name or JS command for adding a chip (enables server-side mode)
  - `on_remove` - Event name or JS command for removing a chip (enables server-side mode)
  - `class` - Additional CSS classes for the container
  - `input_class` - Additional CSS classes for the input field
  - `allow_duplicates` - Whether to allow duplicate values (default: false)

  """
  attr(:id, :string, required: true)
  attr(:name, :string, default: nil)
  attr(:value, :any, default: nil)
  attr(:values, :list, default: [])
  attr(:placeholder, :string, default: "Add value...")
  attr(:suggestions, :list, default: [])
  attr(:on_add, :any, default: nil)
  attr(:on_remove, :any, default: nil)
  attr(:class, :string, default: nil)
  attr(:input_class, :string, default: nil)
  attr(:allow_duplicates, :boolean, default: false)
  attr(:rest, :global)

  @spec chip_input(map()) :: Rendered.t()
  def chip_input(assigns) do
    # Determine mode first
    server_mode = !is_nil(assigns.on_add) || !is_nil(assigns.on_remove)

    # Determine initial values based on mode
    initial_values =
      cond do
        # Server-side mode uses :values
        server_mode -> assigns.values
        # Client-side mode uses :value
        is_list(assigns.value) -> assigns.value
        is_binary(assigns.value) -> String.split(assigns.value, ",", trim: true)
        true -> []
      end

    assigns =
      assigns
      |> assign(:server_mode, server_mode)
      |> assign(:initial_values, initial_values)
      |> assign_new(:has_suggestions, fn -> !Enum.empty?(assigns.suggestions) end)

    ~H"""
    <div
      id={@id}
      class={["chip-input-container", @class]}
      phx-hook="ChipInput"
      data-allow-duplicates={to_string(@allow_duplicates)}
      data-server-mode={to_string(@server_mode)}
      {@rest}
    >
      <%!-- Hidden input for form submission (client-side mode only) --%>
      <%= if !@server_mode && @name do %>
        <input type="hidden" name={@name} value="" data-chip-input-field />
        <%= for value <- @initial_values do %>
          <input type="hidden" name={@name <> "[]"} value={value} data-chip-hidden-value />
        <% end %>
      <% end %>

      <div class={[
        "flex min-h-10 flex-wrap items-center gap-2 rounded-md border border-input bg-background px-3 py-2",
        "focus-within:ring-2 focus-within:ring-ring focus-within:ring-offset-2 focus-within:ring-offset-background"
      ]}>
        <%!-- Existing chips/badges --%>
        <%= for value <- @initial_values do %>
          <.badge variant="secondary" class="gap-1 pl-2 pr-1" data-chip-value={value}>
            {value}
            <button
              type="button"
              data-chip-remove
              phx-click={if @server_mode, do: build_remove_event(@on_remove, @id, value), else: nil}
              class="ml-1 rounded-sm hover:bg-secondary-foreground/20 focus-visible:outline-hidden focus-visible:ring-1 focus-visible:ring-ring"
              aria-label={"Remove #{value}"}
            >
              <.icon name="hero-x-mark" class="size-3" />
            </button>
          </.badge>
        <% end %>

        <%!-- Input field --%>
        <input
          type="text"
          id={"#{@id}-input"}
          data-chip-input
          placeholder={@placeholder}
          class={[
            "flex-1 min-w-[120px] bg-transparent text-sm outline-hidden placeholder:text-muted-foreground",
            "disabled:cursor-not-allowed disabled:opacity-50",
            @input_class
          ]}
          autocomplete="off"
        />
      </div>

      <%!-- Suggestions dropdown (optional) --%>
      <%= if @has_suggestions do %>
        <div data-popover data-state="closed" class="relative">
          <div class="hidden" data-suggestions-trigger></div>
          <div
            data-popover-content
            data-state="closed"
            class={[
              "absolute z-50 mt-1 w-full min-w-[200px] rounded-md border border-border bg-popover p-1 text-popover-foreground shadow-md",
              "data-[state=open]:animate-in data-[state=closed]:animate-out",
              "data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0",
              "data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95",
              "data-[state=closed]:hidden"
            ]}
          >
            <div class="max-h-[200px] overflow-y-auto" role="listbox">
              <%= for suggestion <- @suggestions do %>
                <div
                  role="option"
                  data-suggestion={suggestion}
                  phx-click={build_add_event(@on_add, @id, suggestion)}
                  class={[
                    "flex cursor-pointer items-center rounded-sm px-2 py-1.5 text-sm outline-hidden",
                    "hover:bg-accent hover:text-accent-foreground",
                    "focus-visible:bg-accent focus-visible:text-accent-foreground"
                  ]}
                  tabindex="0"
                >
                  {suggestion}
                </div>
              <% end %>
            </div>
          </div>
        </div>
      <% end %>
    </div>
    """
  end

  # Builds the event for removing a chip
  @spec build_remove_event(any(), String.t(), String.t()) :: JS.t() | String.t()
  defp build_remove_event(nil, _id, _value), do: ""

  defp build_remove_event(event_name, id, value) when is_binary(event_name) do
    JS.push(event_name, value: %{id: id, value: value})
  end

  defp build_remove_event(js_command, _id, _value) when is_struct(js_command, JS) do
    js_command
  end

  # Builds the event for adding a chip
  @spec build_add_event(any(), String.t(), String.t()) :: JS.t() | String.t()
  defp build_add_event(nil, _id, _value), do: ""

  defp build_add_event(event_name, id, value) when is_binary(event_name) do
    event_name
    |> JS.push(value: %{id: id, value: value})
    |> hide_suggestions(id)
  end

  defp build_add_event(js_command, id, _value) when is_struct(js_command, JS) do
    hide_suggestions(js_command, id)
  end

  # Hides the suggestions popover
  @spec hide_suggestions(JS.t(), String.t()) :: JS.t()
  defp hide_suggestions(js, id) do
    js
    |> JS.hide(
      to: "##{id}-suggestions-content",
      transition: {
        "transition ease-in duration-75",
        "opacity-100 scale-100",
        "opacity-0 scale-95"
      }
    )
    |> JS.set_attribute({"data-state", "closed"}, to: "##{id}-suggestions-content")
  end
end

File: assets/js/hooks/chip_input.js
/**
 * Chip Input hook for keyboard interactions
 *
 * Handles:
 * - Enter, Space, Comma: Add new chip from input value
 * - Backspace on empty input: Remove last chip
 * - Escape: Clear input
 * - Show/filter suggestions dropdown as user types
 */
export const ChipInput = {
  mounted() {
    this.input = this.el.querySelector('[data-chip-input]')
    this.allowDuplicates = this.el.dataset.allowDuplicates === 'true'
    this.serverMode = this.el.dataset.serverMode === 'true'
    this.suggestionsContainer = this.el.querySelector('[data-suggestions-trigger]')
    this.suggestionsList = this.el.querySelector('[role="listbox"]')

    if (!this.input) {
      console.warn('ChipInput: Missing input element')
      return
    }

    // Attach remove handlers to existing chips (client-side mode only)
    if (!this.serverMode) {
      this.attachRemoveHandlers()
    }

    // Handle input events for suggestions
    this.handleInput = (e) => {
      if (this.suggestionsList) {
        this.filterSuggestions(e.target.value)
      }
    }

    // Handle focus events to show suggestions
    this.handleFocus = () => {
      if (this.suggestionsContainer && this.suggestionsList) {
        this.showSuggestions()
        this.filterSuggestions(this.input.value)
      }
    }

    // Handle keydown events
    this.handleKeydown = (e) => {
      // Backspace on empty input: Remove last chip
      if (e.key === 'Backspace' && this.input.value === '') {
        e.preventDefault()
        this.removeLastChip()
        return
      }

      const value = this.input.value.trim()

      // Enter, Space, or Comma: Add chip
      if (e.key === 'Enter' || e.key === ' ' || e.key === ',') {
        e.preventDefault()

        if (value) {
          this.addChip(value)
        }
      }
      // Escape: Clear input and hide suggestions
      else if (e.key === 'Escape') {
        e.preventDefault()
        this.input.value = ''
        this.hideSuggestions()
        this.input.blur()
      }
    }

    this.input.addEventListener('keydown', this.handleKeydown)
    this.input.addEventListener('input', this.handleInput)
    this.input.addEventListener('focus', this.handleFocus)

    // Hide suggestions when clicking outside
    this.handleClickOutside = (e) => {
      if (!this.el.contains(e.target)) {
        this.hideSuggestions()
      }
    }
    document.addEventListener('click', this.handleClickOutside)
  },

  showSuggestions() {
    if (this.suggestionsContainer) {
      const popover = this.suggestionsContainer.closest('[data-popover]')
      const content = popover?.querySelector('[data-popover-content]')
      if (popover && content) {
        popover.setAttribute('data-state', 'open')
        content.setAttribute('data-state', 'open')
      }
    }
  },

  hideSuggestions() {
    if (this.suggestionsContainer) {
      const popover = this.suggestionsContainer.closest('[data-popover]')
      const content = popover?.querySelector('[data-popover-content]')
      if (popover && content) {
        popover.setAttribute('data-state', 'closed')
        content.setAttribute('data-state', 'closed')
      }
    }
  },

  filterSuggestions(searchTerm) {
    if (!this.suggestionsList) return

    const options = this.suggestionsList.querySelectorAll('[data-suggestion]')
    const term = searchTerm.toLowerCase()
    let hasVisibleOptions = false

    options.forEach(option => {
      const suggestionText = option.getAttribute('data-suggestion').toLowerCase()
      const matches = suggestionText.includes(term)

      option.style.display = matches ? '' : 'none'
      if (matches) hasVisibleOptions = true
    })

    // Hide suggestions if no matches
    if (!hasVisibleOptions && term) {
      this.hideSuggestions()
    } else if (hasVisibleOptions) {
      this.showSuggestions()
    }
  },

  addChip(value) {
    // Get current chips
    const currentChips = this.getCurrentChips()

    // Check for duplicates if not allowed
    if (!this.allowDuplicates && currentChips.includes(value)) {
      // Clear input and return early
      this.input.value = ''
      return
    }

    if (this.serverMode) {
      // Server-side mode: Push event to LiveView
      this.pushEvent('chip-input:add', {
        id: this.el.id,
        value: value
      })
    } else {
      // Client-side mode: Add chip to DOM directly
      this.addChipToDOM(value)
    }

    // Clear input and hide suggestions
    this.input.value = ''
    this.hideSuggestions()
  },

  removeLastChip() {
    const badges = this.el.querySelectorAll('[data-chip-value]')

    if (badges.length > 0) {
      const lastBadge = badges[badges.length - 1]
      const removeButton = lastBadge.querySelector('[data-chip-remove]')

      if (removeButton) {
        // Trigger the remove event
        removeButton.click()
      }
    }
  },

  getCurrentChips() {
    const badges = this.el.querySelectorAll('[data-chip-value]')
    return Array.from(badges).map(badge => {
      return badge.getAttribute('data-chip-value')
    }).filter(Boolean)
  },

  attachRemoveHandlers() {
    // Find all remove buttons in existing chips and attach click handlers
    const removeButtons = this.el.querySelectorAll('[data-chip-remove]')
    removeButtons.forEach(button => {
      // Skip if already has handler attached
      if (button.dataset.handlerAttached) return

      const badge = button.closest('[data-chip-value]')
      if (badge) {
        const value = badge.getAttribute('data-chip-value')
        button.addEventListener('click', (e) => {
          e.preventDefault()
          this.removeChipFromDOM(value)
        })
        button.dataset.handlerAttached = 'true'
      }
    })
  },

  addChipToDOM(value) {
    // Create badge element matching <.badge variant="secondary" class="gap-1 pl-2 pr-1">
    // Classes must exactly match the server-rendered badge component
    const badge = document.createElement('span')
    badge.className = [
      // Badge base styles (from feedback_status.ex lines 63-67)
      'inline-flex items-center justify-center rounded-md border px-2 py-0.5',
      'text-xs font-medium w-fit whitespace-nowrap shrink-0',
      '[&>svg]:size-3 gap-1 [&>svg]:pointer-events-none',
      'focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px]',
      'transition-[color,box-shadow] overflow-hidden',
      // Badge variant="secondary" styles (from feedback_status.ex lines 85-87)
      'border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/90 [a_&]:hover:bg-secondary/90',
      // Custom class passed to badge in chip_input.ex
      'gap-1 pl-2 pr-1'
    ].join(' ')
    badge.setAttribute('data-chip-value', value)
    badge.textContent = value

    // Create remove button matching <.close_button variant="chip" size="sm">
    // Classes must exactly match the close_button component (from form_input.ex)
    const button = document.createElement('button')
    button.type = 'button'
    button.className = [
      // close_button_classes("chip") (from form_input.ex lines 237-242)
      'ml-1 rounded-sm',
      'hover:bg-secondary-foreground/20',
      'focus-visible:outline-hidden focus-visible:ring-1 focus-visible:ring-ring',
      // close_button_size("sm") (from form_input.ex line 245)
      'p-0.5'
    ].join(' ')
    button.setAttribute('aria-label', `Remove ${value}`)
    button.setAttribute('data-chip-remove', '')
    button.dataset.handlerAttached = 'true' // Mark as having handler

    // Create X icon matching <.icon name="hero-x-mark" class="size-3" />
    // The icon component renders as: <span class="hero-x-mark size-3" />
    const icon = document.createElement('span')
    icon.className = 'hero-x-mark size-3'
    button.appendChild(icon)

    // Add sr-only span for accessibility (matching close_button component)
    const srSpan = document.createElement('span')
    srSpan.className = 'sr-only'
    srSpan.textContent = `Remove ${value}`
    button.appendChild(srSpan)

    // Handle remove click
    button.addEventListener('click', (e) => {
      e.preventDefault()
      this.removeChipFromDOM(value)
    })

    badge.appendChild(button)

    // Insert badge before the input field
    const container = this.input.parentElement
    container.insertBefore(badge, this.input)

    // Add hidden input for form submission
    if (!this.serverMode) {
      this.addHiddenInput(value)
    }
  },

  removeChipFromDOM(value) {
    // Find and remove the badge
    const badges = this.el.querySelectorAll('[data-chip-value]')
    badges.forEach(badge => {
      if (badge.getAttribute('data-chip-value') === value) {
        badge.remove()
      }
    })

    // Remove corresponding hidden input
    if (!this.serverMode) {
      this.removeHiddenInput(value)
    }
  },

  addHiddenInput(value) {
    const container = this.el
    const inputField = container.querySelector('[data-chip-input-field]')

    if (inputField) {
      const hiddenInput = document.createElement('input')
      hiddenInput.type = 'hidden'
      hiddenInput.name = inputField.name.replace(/\[\]$/, '') + '[]'
      hiddenInput.value = value
      hiddenInput.setAttribute('data-chip-hidden-value', '')

      container.insertBefore(hiddenInput, container.firstChild)
    }
  },

  removeHiddenInput(value) {
    const hiddenInputs = this.el.querySelectorAll('[data-chip-hidden-value]')
    hiddenInputs.forEach(input => {
      if (input.value === value) {
        input.remove()
        return
      }
    })
  },

  destroyed() {
    if (this.input) {
      if (this.handleKeydown) {
        this.input.removeEventListener('keydown', this.handleKeydown)
      }
      if (this.handleInput) {
        this.input.removeEventListener('input', this.handleInput)
      }
      if (this.handleFocus) {
        this.input.removeEventListener('focus', this.handleFocus)
      }
    }
    if (this.handleClickOutside) {
      document.removeEventListener('click', this.handleClickOutside)
    }
  }
}
